polkadot/xcm/src/lib.rs

```diff
@@ -451,5 +671,56 @@ pub mod opaque {
 #[test]
 fn conversion_works() {
 	use latest::prelude::*;
-	let _: VersionedMultiAssets = (Here, 1u128).into();
+	let assets: Assets = (Here, 1u128).into();
+	let _: VersionedAssets = assets.into();
+}
+
+#[test]
+fn size_limits() {
+	extern crate std;
+
+	let mut test_failed = false;
+	macro_rules! check_sizes {
+        ($(($kind:ty, $expected:expr),)+) => {
+            $({
+                let s = core::mem::size_of::<$kind>();
+                // Since the types often affect the size of other types in which they're included
+                // it is more convenient to check multiple types at the same time and only fail
+                // the test at the end. For debugging it's also useful to print out all of the sizes,
+                // even if they're within the expected range.
+                if s > $expected {
+                    test_failed = true;
+                    std::eprintln!(
+                        "assertion failed: size of '{}' is {} (which is more than the expected {})",
+                        stringify!($kind),
+                        s,
+                        $expected
+                    );
+                } else {
+                    std::eprintln!(
+                        "type '{}' is of size {} which is within the expected {}",
+                        stringify!($kind),
+                        s,
+                        $expected
+                    );
+                }
+            })+
+        }
+    }
+
+	check_sizes! {
+		(crate::v4::Instruction<()>, 112),
+		(crate::v4::Asset, 80),
+		(crate::v4::Location, 24),
+		(crate::v4::AssetId, 40),
+		(crate::v4::Junctions, 16),
+		(crate::v4::Junction, 88),
+		(crate::v4::Response, 40),
+		(crate::v4::AssetInstance, 40),
+		(crate::v4::NetworkId, 48),
+		(crate::v4::BodyId, 32),
+		(crate::v4::Assets, 24),
+		(crate::v4::BodyPart, 12),
```

Comment:
```suggestion
		(crate::latest::Instruction<()>, 112),
		(crate::latest::Asset, 80),
		(crate::latest::Location, 24),
		(crate::latest::AssetId, 40),
		(crate::latest::Junctions, 16),
		(crate::latest::Junction, 88),
		(crate::latest::Response, 40),
		(crate::latest::AssetInstance, 40),
		(crate::latest::NetworkId, 48),
		(crate::latest::BodyId, 32),
		(crate::latest::Assets, 24),
		(crate::latest::BodyPart, 12),
```

so that we always keep the size in check whenever we create a new XCM version.